processing/src/main/java/org/apache/druid/query/topn/BaseTopNAlgorithm.java

```diff
@@ -315,10 +316,11 @@ public Pair<Integer, Integer> computeStartEnd(int cardinality)
 
       int endIndex = Math.min(ignoreFirstN + keepOnlyN, cardinality);
 
+      final TopNOptimizationInspector topNOptimizationInspector = cursorInspector.getFilteredSegmentInspector();
       if (ignoreAfterThreshold &&
           query.getDimensionsFilter() == null &&
-          !storageAdapter.hasBuiltInFilters() &&
-          query.getIntervals().stream().anyMatch(interval -> interval.contains(storageAdapter.getInterval()))) {
+          (topNOptimizationInspector == null || !topNOptimizationInspector.isFiltered()) &&
```

Comment:
This design seems brittle— it means if a Segment wraps another Segment and filters it, but forgets to implement `TopNOptimizationInspector`, then it'll be treated as non-filtered and the optimization will be allowed, potentially leading to incorrect results. It would be better to design things such that we avoid the optimization in such cases.

Reply:
adjusted and renamed `isFiltered` to `areAllDictionaryIdsPresent` to better reflect usage